### ISSUE TITLE
Can we optimize out preparing?

### ISSUE THREAD

**Issue by mkleehammer** [mkleehammer/pyodbc#214]
Currently pyodbc always prepares statements if there are parameters.  If the statement is identical to the previously prepared, then the previous prepared statement is reused.  However, the SQL Server docs do not recommend this:

[SQL Server 2016 Prepared Execution](https://msdn.microsoft.com/en-us/library/ms131667.aspx)

> Some early ODBC applications used SQLPrepare any time SQLBindParameter was used.
> SQLBindParameter does not require the use of SQLPrepare, it can be used with SQLExecDirect. For
> example, use SQLExecDirect with SQLBindParameter to retrieve the return code or output
> parameters from a stored procedure that is only executed one time. Do not use SQLPrepare with
> SQLBindParameter unless the same statement will be executed multiple times.

Consider a 5.0 API that requires `cursor.prepare(sql)` before using prepared statements to see if we get any performance benefit.

    ps = cursor.prepare(sql)
    row = cursor.execute(ps, var1, var2)

**Comment by gordthompson** [comment]
A similar discussion recently took place [here](https://github.com/Microsoft/mssql-jdbc/issues/166) regarding the SQL Server JDBC driver. The fix that came out of that discussion will call `sp_executesql` (with parameter substitution) the first time around, and automatically fall back to `sp_prepexec` if the prepared statement is executed a second time. However, that evolved from a specific use-case regarding JDBC `PreparedStatement` objects where an explicit "prepare" is implied by the `prepareStatement` method that is used to create the object.

An explicit `.prepare()` method for a pyodbc Cursor would actually make it more like the .NET `OdbcCommand` objects that have an explicit `.Prepare()` method. (ADO.NET decided to go with a single `...Command` object instead of separate `Statement` and `PreparedStatement` objects as JDBC did.)

**Comment by TobiasSQL** [comment]
Thanks @gordthompson, I agree, for SQL Server it sounds like following the pattern that we used for JDBC should work well here too. If prepare is required in pyodbc for the use of parameters they sound analog to each other. 

BTW, for .NET we are discussing following the pattern even for an explicit call to Prepare (even though prepare is not needed for parameter binding). The thought is that the combination of a) it having negligible performance impact to call sp_executesql on the first call and next falling back to sp_prepexec, and b) there is a fairly high risk that developers call Prepare() even when they don't need it which would hurt performance by introducing additional roundtrips or unnecessary memory consumption on the server.

Thoughts?

/Tobias (SQL Server database engine)

**Comment by TobiasSQL** [comment]
+@meet-bhagdev who is looking out for Python in SQL Server :-)

**Comment by gordthompson** [comment]
@TobiasSQL 

While I can appreciate how a subtle performance optimization could accrue significant benefits as we scale up, I also have some mild reservations based on the [Principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment): If I called `.prepare` followed by `.execute` and then looked at SQL Profiler to see the result I would expect to see `sp_prepare` followed by `sp_execute`. Since most of the Python data-access layers that follow the [DB-API](https://www.python.org/dev/peps/pep-0249/) don't have an explicit `.prepare` method I was able to accept repeated `sp_prepexec` calls as a reasonable alternative, but an explicit `.prepare` resulting in an initial `sp_executesql` and *then* a call to `sp_prepexec` (for the second .execute) might leave me scratching my head.

**Comment by TobiasSQL** [comment]
That makes sense, what about sp_unprepare? I get this from a "want to understand what happens" POV but I am wondering what would actually help most apps get the best perf experience :-)

**Comment by mkleehammer** [comment]
So we are all on the same page, what happens today is that any statement with parameters is prepared and parameters are bound.  If the next call to execute is for the same SQL statement (by identity, not content at this time), then the previously prepared statement is used.  If I understood the Java discussion correctly, that is what was proposed.  Did I miss anything?

I believe the current issue is stating that *binding* parameters is causing new execution plans to be created because the parameters have different lengths.  For example, if the same prepared statement is used for a 10-character string and then for a 20-character string, it would be bound as SQL_WVARCHAR(10) the first time and SQL_WVARCHAR(20) the second time.  I have not verified that this would actually cause a different execution plan.

I believe the prepare part works.  I have not looked at the binding yet.

Did I misunderstand any of these?

**Comment by mkleehammer** [comment]
Please ignore the 2nd half of my previous comment - I followed a link from email and thought this was the related issue #213.

Someone above mentioned that pyodbc might require a prepare for binding - it is doing that now because I originally thought ODBC required it but the SQL Server docs say otherwise.  We should be able to call SQLExecuteW and bind parameters.  I suspect many drivers will not, but we can default to this and use a flag for those that don't.

One issue to be aware of is ODBC only allows one prepared statement per HSTMT, which is what we represent with a cursor.  If you need to switch between two prepared statements you'd want to create two cursors.

**Comment by gordthompson** [comment]
re: "what about sp_unprepare?"

To be honest, I haven't worried too much about sp_unprepare. Over the last while I had noticed that several reputable implementations did not send an sp_unprepare for every sp_prepexec they sent (e.g., pyodbc ref: [here](http://stackoverflow.com/a/40156617/2144390), and .NET SqlClient ref: [here](https://github.com/Microsoft/mssql-jdbc/issues/166#issuecomment-284248272)). I suspected that they were leaving un-unprepared entries in a list somewhere on the server (and consuming numeric handles) but I took it on faith that the lack of cleanup wasn't likely to cause significant problems.

re: binding without preparing

If it can be done then I'd definitely support that. Again, it would be like SqlClient where .Execute without .Prepare does sp_executesql (with binding if necessary) and multiple .Executes on the same SqlCommand after a .Prepare will do an sp_prepexec followed by one or more sp_execute calls. That way, 

- an explicit .Prepare could still be used to reduce traffic by not sending the command text each time (which can be a significant bottleneck for .executemany as in #120, especially if the command text is lengthy), but

- without an explicit .Prepare we could get the more direct sp_executesql approach (instead of sp_prepexec) and still benefit from the execution plan cache without the prepared statement "cleanup" issues mentioned above.

**Comment by TobiasSQL** [comment]
@meet-bhagdev and myself did a bit of playing around with pyodbc and this is what we found out, at least as related to SQL Server :-) Hopefully it can help.

For more background on prepared statements in SQL Server you can take a look at [my first reply](https://github.com/Microsoft/mssql-jdbc/issues/166#issuecomment-284490675) to the JDBC issue that @gordthompson mentioned. 

After testing pyodbc a bit with SQL Server it seems like pyodbc never re-uses a server-side prepared statement handle (please correct if wrong). I.e. you either do cursor.execute("my SQL code", ..) with or without params. There seems to be no way to re-issue the same "prepared statement" without sending the SQL code to the server again (something like cursor.executePrepared(statementHandle, <params>) if that is not what cusor.executeMany does...).

Right now the following seems to be the case:
1. cursor.execute("SELECT ...") - Execute without parameters. Calls ODBC SQLExecDirect which for SQL Server will generate a language event/SQL batch call. 
2. cursor.execute("SELECT ... WHERE col = ?", 42) - Execute with parameters. Calls ODBC SQLPrepare to bind parameters followed by SQLExecute at which time the SQL Server driver will generate a TDS RPC request to sp_prepexec in turn preparing a handle that can be re-used on the server. 
3. On connection close ODBC SQLFreeStmt seems to be called to unprepare the last prepared statement (sp_unprepare for SQL Server).

Here is my SQL Server trace:
![pyodbc_screenshot](https://cloud.githubusercontent.com/assets/1648796/23973586/06b97378-0994-11e7-9041-fcc0975d1f2b.png)

Looking at these results it seems like:
1. Is great :-)
2. Should be changed to call SQLExecDirect instead of SQLExecute, at least for SQL Server. You still call SQLPrepare first since that should not issue a call to the server and I think that is required for parameter binding as mentioned earlier in the thread. 
3. If other RDBMSs require the use of SQLPrepare, SQLFreeStmt would need to be called after each statement.

Thanks!
/Tobias

**Comment by Lexcon** [comment]
The last time I checked the odbc driver of sql server using a network sniffer, it didn't actually use prepared queries at all, it kept sending the entire query over the wire. Maybe msft considers it a non-optimisation.

That said, for other db systems: it sounds like a keepprepared parameter might be useful when running a query. Consider this alternating sequence of queries:

- fetch user
- update user
- fetch next user
- update that user

Because of the alternation, the queries plan will not remain available since any prepared query gets dropped the moment the next query plan needs to be prepared.

To facilitate a 'keepprepared' parameter, maybe it would be smart to have a dict like structure that keeps hashes of a query mapped to the compiled plan, then before running the query first checks if a prepared query already exists.

I think that's how sql server does it server side anyway, so that might be duping existing funtionality: I think preparing a query is just a lookup on a highly optimized database server system.

Anyway 'keepprepared' should maybe have a time value in seconds to avoid memory buildup in long running programs. That makes it a bit bloated. It has some risk of being a senseless optimisation one can only know that after there's a benchmark, so after it's created.

**Comment by mkleehammer** [comment]
@gordthompson 

I've written up some thoughts on [parameter binding](https://github.com/mkleehammer/pyodbc/wiki/Binding-Parameters).  I'm looking for input from anyone interested.

In particular, the API looks very messy but I'm looking to start a discussion.

**Comment by bkline** [comment]
## Overview

@mkleehammer I've done some analysis of the options on the [Binding Parameters wiki page](https://github.com/mkleehammer/pyodbc/wiki/Binding-Parameters). I created a proof-of-concept implementation for the first option and from what I can tell it does not appear to offer any performance benefits. In fact, running with the patched pyodbc without using any of the new properties, insertion of rows with parameters takes at least twice as long as with the current released pyodbc. If I enable caching of the information needed for binding I can get the performance close to what it is for the unpatched pyodbc. The same is true if I suppress the calls to `SQLPrepareW()` and `SQLDescribeParam()` by setting `var_binding_length`.

I did some deeper digging and as it turns out, `SQLPrepareW()` is not a very expensive call. For one thing, in the normal course of things, when an application has a `Cursor` which is pumping a sequence of rows into a table, pyodbc only calls `SQLPrepareW()` for the first row, recognizing with subsequent rows that the SQL query string is the same object as the one used for the first row. Even if that optimization is broken (for example, by using a dynamic f-string with interpolations for the `INSERT` string), 20,000 calls to `SQLPrepareW()` only take about a second and a half on my laptop, testing against a back end on a server over the network (well under 1/10 of a millisecond per call).

## Profiling

I created a script to time inserting 1,000 rows into a 5-column table, rolling back the transaction, and then doing it again, for 10 iterations. I ran the script twice for each scenario, with ODBC trace logging enabled into a per-scenario log file. I then wrote a parser to extract and accumulate the timings for each ODBC call from those logs. Results are provided below the script.

### Script

```python
"""Show effect of pyodbc ticket #214 optimizations"""

from argparse import ArgumentParser
from datetime import date, datetime
from decimal import Decimal
from pyodbc import connect, ProgrammingError

COMMENT = "yada yada " * 256
INSERT = f"INSERT INTO bench_0214 (a,b,c,d,e) -- {COMMENT}\nVALUES (?,?,?,?,?)"


def main():
    """Top-level driver"""
    parser = ArgumentParser()
    parser.add_argument("connstr")
    parser.add_argument("--iterations", "-i", type=int, default=10)
    parser.add_argument("--rows", "-r", type=int, default=1000)
    parser.add_argument("--collen", "-l", type=int)
    parser.add_argument("--cachesize", "-c", type=int)
    opts = parser.parse_args()
    conn = connect(opts.connstr)
    if opts.collen:
        if not hasattr(conn, "var_binding_length"):
            print("var_binding_length not supported in this build")
            return
        conn.var_binding_length = opts.collen
    if opts.cachesize:
        if not hasattr(conn, "bindinfo_cache_size"):
            print("bindinfo_cache_size not supported in this build")
            return
        conn.bindinfo_cache_size = opts.cachesize
    cursor = conn.cursor()
    try:
        cursor.execute("DROP TABLE bench_0214")
    except ProgrammingError:
        pass
    cursor.execute("""\
        CREATE TABLE bench_0214 (
            a INTEGER,
            b VARCHAR(256),
            c DECIMAL(5,2),
            d DATE,
            e FLOAT
        )
    """)
    conn.commit()
    data = (
        (1, "abcd" * 0, Decimal("123.45"), date(2001, 1, 1), 3.14),
        (2, "abcd" * 1, Decimal("234.56"), date(2002, 2, 2), 3.141),
        (3, "abcd" * 2, Decimal("345.67"), date(2003, 3, 3), 3.1416),
        (4, "abcd" * 3, Decimal("456.78"), date(2004, 4, 4), 3.14159),
        (5, "abcd" * 4, Decimal("567.89"), date(2005, 5, 5), 3.141593),
        (6, "abcd" * 5, Decimal("678.90"), date(2006, 6, 6), 3.1415927),
        (7, "abcd" * 6, Decimal("789.01"), date(2007, 7, 7), 3.14159265),
        (8, "abcd" * 7, Decimal("890.12"), date(2008, 8, 8), 3.141592654),
        (9, "abcd" * 8, Decimal("901.23"), date(2009, 9, 9), 3.1415926535),
    )
    start = datetime.now()
    iteration = 0
    while iteration < opts.iterations:
        inserted = 0
        while inserted < opts.rows:
            values = data[inserted % len(data)]
            cursor.execute(INSERT, values)
            inserted += 1
        conn.rollback()
        iteration += 1
    elapsed = datetime.now() - start
    print(
        f"elapsed={elapsed} cachesize={opts.cachesize} collen={opts.collen} "
        f"iterations={opts.iterations} rows={opts.rows}"
    )


if __name__ == "__main__":
    main()
```

### Unpatched pyodbc

This is what the script gets for the pair of runs using the currently released pyodbc (5.3.0).

Seconds in Call | # Calls | Per Call (ms) | Function
---: | ---: | ---: | ---
30.762 | 20,000 | 1.538 | SQLExecute
7.279 | 100,000 | 0.073 | SQLBindParameter
2.957 | 40,014 | 0.074 | SQLFreeStmt
1.481 | 20,004 | 0.074 | SQLRowCount
1.444 | 20,004 | 0.072 | SQLNumResultCols
0.318 | 2 | 159.241 | SQLDriverConnectW
0.116 | 24 | 4.836 | SQLEndTran
0.018 | 4 | 4.402 | SQLExecDirectW
0.017 | 8 | 2.153 | SQLGetTypeInfo
0.005 | 2 | 2.298 | SQLSetConnectAttr
0.003 | 14 | 0.213 | SQLFreeHandle
0.001 | 12 | 0.044 | SQLAllocHandle
0.000 | 2 | 0.233 | SQLGetEnvAttr
0.000 | 8 | 0.038 | SQLFetch
0.000 | 8 | 0.037 | SQLGetData
0.000 | 2 | 0.098 | SQLDisconnect
0.000 | 4 | 0.040 | SQLGetInfo
0.000 | 2 | 0.046 | SQLPrepareW
0.000 | 2 | 0.035 | SQLNumParams
0.000 | 2 | 0.029 | SQLSetEnvAttr

### Dynamic command string

With the optimization to bypass extra calls to `SQLPrepareW()` suppressed, those calls don't have a serious impact on performance, only increasing the wall clock time from 59 to 60 seconds (see the table in the _Driver vs. pyodbc_ section below).

Seconds in Call | # Calls | Per Call (ms) | Function
---: | ---: | ---: | ---
30.231 | 20,000 | 1.512 | SQLExecute
6.308 | 100,000 | 0.063 | SQLBindParameter
2.573 | 40,014 | 0.064 | SQLFreeStmt
1.534 | 20,000 | 0.077 | SQLPrepareW
1.277 | 20,004 | 0.064 | SQLRowCount
1.265 | 20,000 | 0.063 | SQLNumParams
1.262 | 20,004 | 0.063 | SQLNumResultCols
0.322 | 2 | 160.935 | SQLDriverConnectW
0.112 | 24 | 4.655 | SQLEndTran
0.016 | 8 | 1.970 | SQLGetTypeInfo
0.013 | 4 | 3.195 | SQLExecDirectW
0.003 | 2 | 1.400 | SQLSetConnectAttr
0.003 | 14 | 0.193 | SQLFreeHandle
0.001 | 12 | 0.048 | SQLAllocHandle
0.000 | 8 | 0.061 | SQLGetData
0.000 | 8 | 0.040 | SQLFetch
0.000 | 4 | 0.040 | SQLGetInfo
0.000 | 2 | 0.077 | SQLDisconnect
0.000 | 2 | 0.036 | SQLGetEnvAttr
0.000 | 2 | 0.034 | SQLSetEnvAttr

### Patched pyodbc, default behavior

These are the numbers for the first option without caching enabled or the new `var_binding_length` property set. In this scenario, the calls to `SQLDescribeParam()` are the elephant in the room.

Seconds in Call | # Calls | Per Call (ms) | Function
---: | ---: | ---: | ---
57.057 | 100,000 | 0.571 | SQLDescribeParam
26.725 | 20,000 | 1.336 | SQLExecute
8.701 | 100,000 | 0.087 | SQLBindParameter
3.330 | 40,014 | 0.083 | SQLFreeStmt
1.656 | 20,004 | 0.083 | SQLNumResultCols
1.647 | 20,004 | 0.082 | SQLRowCount
0.301 | 2 | 150.617 | SQLDriverConnectW
0.137 | 24 | 5.712 | SQLEndTran
0.018 | 4 | 4.393 | SQLExecDirectW
0.016 | 8 | 1.984 | SQLGetTypeInfo
0.003 | 14 | 0.221 | SQLFreeHandle
0.003 | 2 | 1.379 | SQLSetConnectAttr
0.002 | 4 | 0.573 | SQLGetInfo
0.001 | 12 | 0.045 | SQLAllocHandle
0.000 | 2 | 0.241 | SQLGetEnvAttr
0.000 | 8 | 0.046 | SQLGetData
0.000 | 8 | 0.037 | SQLFetch
0.000 | 2 | 0.113 | SQLDisconnect
0.000 | 2 | 0.040 | SQLPrepareW
0.000 | 2 | 0.036 | SQLNumParams
0.000 | 2 | 0.030 | SQLSetEnvAttr

### Caching enabled

Turning on caching brings the performance almost (but not quite) up to the level of what we get with the unpatched pyodbc.

Seconds in Call | # Calls | Per Call (ms) | Function
---: | ---: | ---: | ---
31.128 | 20,000 | 1.556 | SQLExecute
7.331 | 100,000 | 0.073 | SQLBindParameter
3.003 | 40,014 | 0.075 | SQLFreeStmt
1.465 | 20,004 | 0.073 | SQLRowCount
1.456 | 20,004 | 0.073 | SQLNumResultCols
0.305 | 2 | 152.741 | SQLDriverConnectW
0.125 | 24 | 5.205 | SQLEndTran
0.017 | 4 | 4.154 | SQLExecDirectW
0.016 | 8 | 2.032 | SQLGetTypeInfo
0.005 | 10 | 0.526 | SQLDescribeParam
0.003 | 14 | 0.238 | SQLFreeHandle
0.002 | 2 | 1.232 | SQLSetConnectAttr
0.001 | 12 | 0.048 | SQLAllocHandle
0.000 | 2 | 0.247 | SQLGetEnvAttr
0.000 | 8 | 0.047 | SQLGetData
0.000 | 8 | 0.045 | SQLFetch
0.000 | 2 | 0.117 | SQLDisconnect
0.000 | 4 | 0.042 | SQLGetInfo
0.000 | 2 | 0.046 | SQLSetEnvAttr
0.000 | 2 | 0.041 | SQLPrepareW
0.000 | 2 | 0.033 | SQLNumParams

### Binding information derived from values

With the `var_binding_length` property set, the script can bypass the calls to `SQLPrepareW()` and (more significantly for performance) `SQLDescribeParam()`.

Seconds in Call | # Calls | Per Call (ms) | Function
---: | ---: | ---: | ---
30.739 | 20,004 | 1.537 | SQLExecDirectW
6.777 | 100,000 | 0.068 | SQLBindParameter
2.808 | 40,014 | 0.070 | SQLFreeStmt
1.390 | 20,004 | 0.069 | SQLRowCount
1.366 | 20,004 | 0.068 | SQLNumResultCols
0.331 | 2 | 165.741 | SQLDriverConnectW
0.156 | 24 | 6.499 | SQLEndTran
0.017 | 8 | 2.129 | SQLGetTypeInfo
0.003 | 2 | 1.277 | SQLSetConnectAttr
0.001 | 14 | 0.069 | SQLFreeHandle
0.001 | 12 | 0.050 | SQLAllocHandle
0.000 | 2 | 0.233 | SQLGetEnvAttr
0.000 | 2 | 0.206 | SQLSetEnvAttr
0.000 | 8 | 0.051 | SQLGetData
0.000 | 8 | 0.041 | SQLFetch
0.000 | 2 | 0.129 | SQLDisconnect
0.000 | 4 | 0.039 | SQLGetInfo

### Driver _vs._ pyodbc 

I wanted to get the broader context for the numbers above, which represent time elapsed between when we call an ODBC API function and when that function returns. Across all the scenarios those calls consistently account for about three quarters of the total elapsed time for the script runs, the remaining quarter representing processing of pyodbc's own code. As you would expect, that proportion increases slightly in favor of the ODBC calls with the addition of the calls to `SQLDescribeParam()` in the third scenario.

Scenario | Elapsed | ODBC Calls | Pct | pyodbc | Pct
--- | ---: | ---: | ---: | ---: | ---:
Unpatched pyodbc | 59.479 | 44.401 | 74.7% | 15.078 | 25.3
Dynamic SQL query string | 60.391 | 44.920 | 74.4% | 15.471 | 25.6
Patched pyodbc using defaults | 128.348 | 99.597 | 77.6% | 28.751 | 22.4
Patched pyodbc, cache enabled | 60.089 | 44.857 | 74.7% | 15.232 | 25.3
Patched pyodbc, var_binding_length set | 57.858 | 43.589 | 75.3% | 14.269 | 24.7

## Option 2

I didn't implement anything for the second option, as it would involve behavior changes which are not opt-in, but would kick in automatically, potentially breaking existing code (for example, our own tests which loop through different schemas using the same table, which would result in the wrong binding information being pulled from the cache and applied).

## Other notes

There are already quite a few user-controllable knobs which interact with each other, and the prototype I implemented introduced even more. One obvious overlap would involve the existing `inputsizes` property and the proposed `var_binding_length` property. The new property can be tricky: if you use it to improve performance for `VARCHAR` columns you can easily break smaller `CHAR` columns (or smaller `VARCHAR` columns). Another area of overlap has two separate mechanisms trying to solve the same problem with `none_binding` and `null_binary`. Also, it's awkward that `var_binding_length` is overloaded to answer two questions in the same property ("should we compute binding information without using `SQLDescribeParam()`?" and "should we round up column lengths in binding? and if so by how much?"). If we were looking at measurable performance gains, we'd be weighing those benefits against the cost of additional complexity. As it is, however, there doesn't seem to be a whole lot on the plus side of the ledger.

At this point I'm not planning on submitting a PR, but my prototype for the first option can be reviewed at https://github.com/bkline/pyodbc/tree/214-optimize-bind-info. Of course, it's possible that I have misunderstood how the proposed solution was intended to work, or have overlooked a way to implement it more efficiently. Feedback is always welcome.

**TL;DR**: this seems like it might be a case of premature optimization. 🤔
